File: TypeScript/app/system.ts
import { IAuthorizer } from "./IAuthorizer";
import { IMailService } from "./IMailService";

export default class System {
  public static readonly errorInvalidUserName: 0;
  private authorizer: IAuthorizer | null;
  private mailService: IMailService | null;
  private _loginCount: number = 0;

  constructor(
    authorizer: IAuthorizer | null,
    mailService: IMailService | null
  ) {
    this.authorizer = authorizer;
    this.mailService = mailService;
  }

  public loginCount(): number {
    return this._loginCount;
  }

  public performLogin(userName: string, password: string): void {
    if (!userName) {
      this.mailService?.sendSystemFailureMail(System.errorInvalidUserName);
      return;
    }

    if (!this.authorizer?.authorize(userName, password)) {
      this.mailService?.sendLoginFailureMail(userName);
      return;
    }

    this._loginCount++;
  }
}

File: CSharpCore/CSharpCore/System.cs
using System;

namespace CSharpCore
{
    public class System
    {
        public static readonly int ErrorInvalidUserName = 5;
        private readonly IAuthorizer authorizer;
        private IMailService mailService;
        private int loginCount = 0;

        public System(IAuthorizer authorizer, IMailService mailService)
        {
            this.authorizer = authorizer;
            this.mailService = mailService;
        }

        public int LoginCount()
        {
            return loginCount;
        }

        public void PerformLogin(string userName, string password)
        {
            if (String.IsNullOrEmpty(userName))
            {
                mailService.SendSystemFailureMail(System.ErrorInvalidUserName);
                return;
            }

            if (!authorizer.Authorize(userName, password))
            {
                mailService.SendLoginFailureMail(userName);
                return;
            }

            loginCount++;
        }
    }
}
